File: reactivit/reactivitapp/src/main/java/net/medrag/reactivit/reactivitapp/filter/HeadersFilter.java
package net.medrag.reactivit.reactivitapp.filter;

import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import org.springframework.web.server.WebFilter;
import org.springframework.web.server.WebFilterChain;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Optional;

@Slf4j
@Component
public class HeadersFilter implements WebFilter {

    public static final String REQUEST_ID = "rquid";
    public static final String NAME = "name";
    public static final String SNAME = "sname";
    public static final String SENIORITY = "seniority";

    @Override
    public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
        final var requestIdOpt = getRequestId(exchange.getRequest().getHeaders());
        if (requestIdOpt.isPresent()) {
            final var requestId = requestIdOpt.get();
            exchange.getResponse().getHeaders().add(REQUEST_ID, requestId);
            return chain.filter(exchange).contextWrite(ctx -> ctx.put(REQUEST_ID, requestId));
        } else {
            return chain.filter(exchange);
        }
    }

    private Optional<String> getRequestId(HttpHeaders headers) {
        List<String> requestIdHeaders = headers.get(REQUEST_ID);
        return requestIdHeaders == null || requestIdHeaders.isEmpty()
            ? Optional.empty()
            : Optional.ofNullable(requestIdHeaders.get(0));
    }
}

File: ktapp-parent/ktapp/src/main/kotlin/net/medrag/ktapp/filter/HeadersFilter.kt
package net.medrag.ktapp.filter

import net.medrag.ktapp.context.ProcessContext
import net.medrag.ktapp.context.ProcessContextParameter
import net.medrag.ktapp.context.addContext
import net.medrag.ktapp.ext.header
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import org.springframework.web.server.WebFilter
import org.springframework.web.server.WebFilterChain
import reactor.core.publisher.Mono
import reactor.util.context.ContextView

/**
 * Filter extracts http headers from [ServerWebExchange] and puts it into the Reactor context as a part of [ProcessContext].
 * Later it can be retrieved from the [ContextView].
 */
@Component
class HeadersFilter : WebFilter {

    override fun filter(exchange: ServerWebExchange, chain: WebFilterChain): Mono<Void> {
        val requestId: String? = exchange.header(RequestHeaders.REQUEST_ID)
        val requesterUid: String? = exchange.header(RequestHeaders.REQUESTER_UID)
        val someData: String? = exchange.header(RequestHeaders.SOME_DATA)

        var processContext = ProcessContext.of(
            ProcessContextParameter.REQUEST_HEADERS to mapOf(
                RequestHeaders.REQUEST_ID to requestId,
                RequestHeaders.REQUESTER_UID to requesterUid,
                RequestHeaders.SOME_DATA to someData,
            )
        )
        if (!requestId.isNullOrEmpty()) {
            exchange.response.headers.add(RequestHeaders.REQUEST_ID, requestId)
            processContext = processContext.enrich(ProcessContextParameter.REQUEST_ID, requestId)
        }
        if (!requesterUid.isNullOrEmpty()) {
            exchange.response.headers.add(RequestHeaders.REQUESTER_UID, requesterUid)
            processContext = processContext.enrich(ProcessContextParameter.REQUESTER_UID, requesterUid)
        }
        if (!someData.isNullOrEmpty()) {
            exchange.response.headers.add(RequestHeaders.SOME_DATA, someData)
            processContext = processContext.enrich(ProcessContextParameter.SOME_DATA, someData)
        }

        println(processContext)
        return chain
            .filter(exchange)
            .addContext(processContext)
    }
}